README.txt
Introduction
============

``repoze.who.plugins.openid`` is a plugin for the `repoze.who framework <http://static.repoze.org/whodocs/>`_
enabling `OpenID <http://openid.net>`_ logins.

For more information read the `documentation <http://quantumcore.org/docs/repoze.who.plugins.openid>`_ 
or check out the `source code <http://code.google.com/p/repozeopenid/>`_.




docs/HISTORY.txt
Changelog
=========


0.5.3 - TBD
-----------

* added webob to requirements in setup.py



0.5.2 - September 25th 2009
---------------------------

* added MANIFEST.in to make egg work again 



0.5.1 - September 25th 2009
---------------------------

* fixed setup.py (removed doubles test suite)



0.5 - March 15th 2009
---------------------

* Initial release





repoze/who/plugins/openid/tests/consumer.py
"""

This code is copied from plone.openid 
http://svn.plone.org/svn/plone/plone.openid/trunk/plone/openid/tests/consumer.py

"""
from openid.consumer.consumer import FAILURE, SUCCESS

class MockAuthRequest:
    """Amock OpenID AuthRequest.
    """
    def __init__(self, status=None, identity_url=None, message=None):
        self.status=status
        self.identity_url=identity_url
        self.message=message
        self.extensions = []


    def redirectURL(self, trust_root, return_to):
        return "http://someopenidprovider.com/somewhere"

    def getDisplayIdentifier(self):
        return "http://foobar.com"
    def addExtension( self, ext_message ):
        self.extensions.append( ext_message )

class MockConsumer:
    """A mock OpenID consumerclass.
    """

    def begin(self, identity):
        self.identity=identity
        return MockAuthRequest()

    def complete(self, credentials, current_url):
        status=SUCCESS
        message="authentication completed succesfully"
        
        if credentials.has_key("openid.identity") and credentials["openid.identity"] == "":
            # if the python openid is passed an identity of an empty string
            # an IndexError is raised in the depths of its XRI identification
            # see: http://www.oasis-open.org/committees/tc_home.php?wg_abbrev=xri

            # an empty string is common when the submit button of the
            # openid login is clicked prior to providing an identity url
            # we simulate openid's response here in our mock object
            message="invalid identity"
            status=FAILURE
        else:
            for field in [ "nonce", "openid.identity",
                    "openid.assoc_handle", "openid.return_to", "openid.signed",
                    "openid.sig", "openid.invalidate_handle", "openid.mode"]:
                if field not in credentials:
                    message="field missing"
                    status=FAILURE
                

        return MockAuthRequest(status=status,
                                message=message,
                                identity_url=credentials["openid.identity"])

def get_consumer(environ):
    return MockConsumer()

def patch_plugin(plugin):
    plugin.get_consumer = get_consumer
    return plugin
    




repoze/who/plugins/openid/tests/test_challenge.py
import unittest
import StringIO

from repoze.who.tests import encode_multipart_formdata
from StringIO import StringIO

from repoze.who.plugins.openid.classifiers import openid_challenge_decider
from repoze.who.plugins.openid.identification import OpenIdIdentificationPlugin

from consumer import patch_plugin

class ChallengeTest(unittest.TestCase):
    """test the challenge plugin"""

    def setUp(self):
        self.server_response={
                "openid.mode"              : "id_res",
                "nonce"                    : "nonce",
                "openid.identity"          : "http://repoze.myopenid.com",
                "openid.assoc_handle"      : "assoc_handle",
                "openid.return_to"         : "return_to",
                "openid.signed"            : "signed",
                "openid.sig"               : "sig",
                "openid.invalidate_handle" : "invalidate_handle",
            }
        self.plugin = patch_plugin(OpenIdIdentificationPlugin(
                                        store = None,
                                        openid_field = 'repoze.whoplugins.openid.openid',
                                        error_field = '',
                                        store_file_path='',
                                        session_name = '',
                                        login_handler_path = '/login',
                                        logout_handler_path = '',
                                        login_form_url = '/login_form',
                                        logged_in_url = '',
                                        logged_out_url = '',
                                        came_from_field = 'came_from',
                                        rememberer_name = ''
                                        )
                                    )

        environ = {'wsgi.input':'',
            'wsgi.url_scheme': 'http',
            'SERVER_NAME': 'localhost',
            'SERVER_PORT': '8080',
            'CONTENT_TYPE':'text/html',
            'CONTENT_LENGTH':0,
            'REQUEST_METHOD':'POST',
            'PATH_INFO': '/protected',
            'QUERY_STRING':'',
        }

        class DummyLogger:
            warnings = []
            debugs = []
            infos = []
            def warn(self, msg):
                self.warnings.append(msg)
            def debug(self, msg):
                self.debugs.append(msg)
            def info(self, msg):
                self.infos.append(msg)
        logger = environ['repoze.who.logger'] = DummyLogger()

        self.environ=environ

    def tearDown(self):
        pass

    def test_challenge_decider(self):
        """test challenge decider"""

        environ = self.environ
        environ['repoze.whoplugins.openid.openid'] = 'foobar.com'

        # decider takes environ, status, headers
        self.assertEqual(openid_challenge_decider(environ, '200 Ok', {}), True)
        self.assertEqual(openid_challenge_decider({}, '401 Unauthorized', {}), True)
        self.assertEqual(openid_challenge_decider({}, '200 Ok', {}), False)

    def test_challenge_redirect(self):
        """check if the challenge plugin works if given an openid"""

        # create a form POST response as if we would post the openid
        fields = [('repoze.whoplugins.openid.openid','foobar.com')]
        content_type, body = encode_multipart_formdata(fields)

        environ = self.environ
        environ['wsgi.input'] = StringIO(body)
        environ['REQUEST_METHOD'] = 'POST'
        environ['CONTENT_LENGTH'] = len(body)
        environ['CONTENT_TYPE'] = content_type

        # in this case the plugin has to redirect to the openid provider
        # faked by MockConsumer in this case
        res = self.plugin.challenge(environ, '200 Ok', {}, {})
        self.assertEqual(res.location,'http://someopenidprovider.com/somewhere')
        self.assertEqual(res.status,'302 Found')

    def test_challenge_show_login_form(self):
        """test if the challenge plugin redirects to the login form"""

        res = self.plugin.challenge(self.environ, '200 Ok', {}, {})
        self.assertEqual(res.location,'/login_form?came_from=http://localhost:8080/protected')
        self.assertEqual(res.status,'302 Found')

    def test_login_form_send(self):
        """test if the login form data is received and the environment set correctly"""
        fields = [('repoze.whoplugins.openid.openid','foobar.com')]
        content_type, body = encode_multipart_formdata(fields)

        environ = self.environ
        environ['wsgi.input'] = StringIO(body)
        environ['REQUEST_METHOD'] = 'POST'
        environ['CONTENT_LENGTH'] = len(body)
        environ['CONTENT_TYPE'] = content_type
        environ['PATH_INFO'] = '/login'

        identity = self.plugin.identify(environ)
        self.assertEqual(environ.get('repoze.whoplugins.openid.openid',None), 'foobar.com')

    def test_complete_openid_request(self):
        """test if the openid request completes"""

        environ = self.environ
        environ['PATH_INFO'] = '/login'

        fields = self.server_response.items()
        content_type, body = encode_multipart_formdata(fields)
        environ['wsgi.input'] = StringIO(body)
        environ['REQUEST_METHOD'] = 'POST'
        environ['CONTENT_LENGTH'] = len(body)
        environ['CONTENT_TYPE'] = content_type
        
        identity = self.plugin.identify(environ)
        self.assertEqual(identity['repoze.who.plugins.openid.userid'],'http://repoze.myopenid.com')

    def test_incomplete_openid_request(self):
        """test if the openid request fails with a wrong identity"""

        environ = self.environ
        environ['PATH_INFO'] = '/login'

        sresp = self.server_response
        sresp['openid.identity'] = ''

        fields = sresp.items()
        content_type, body = encode_multipart_formdata(fields)
        environ['wsgi.input'] = StringIO(body)
        environ['REQUEST_METHOD'] = 'POST'
        environ['CONTENT_LENGTH'] = len(body)
        environ['CONTENT_TYPE'] = content_type
        
        identity = self.plugin.identify(environ)
        self.assertEqual(identity.get('repoze.who.plugins.openid.userid',None),None)

    def test_authenticate(self):
        """test if the authentication plugin works as well"""
        environ = self.environ
        identity = {'repoze.who.plugins.openid.userid' : 'http://foobar.com'}
        res = self.plugin.authenticate(environ, identity)
        self.assertEqual(res, 'http://foobar.com')


repoze/who/plugins/openid/__init__.py
from identification import OpenIdIdentificationPlugin

def make_identification_plugin(store='mem',
                openid_field = "openid",
                session_name = None,
                login_handler_path = None,
                logout_handler_path = None,
                login_form_url = None,
                error_field = 'error',
                logged_in_url = None,
                logged_out_url = None,
                came_from_field = 'came_from',
                store_file_path='',
                rememberer_name = None,
                sql_associations_table = '',
                sql_nonces_table = '',
                sql_connstring = ''):
    if store not in (u'mem',u'file',u'sql'):
        raise ValueError("store needs to be 'mem', 'sql' or 'file'")
    if login_form_url is None:
        raise ValueError("login_form_url needs to be given")
    if rememberer_name is None:
        raise ValueError("rememberer_name needs to be given")
    if login_handler_path is None:
        raise ValueError("login_handler_path needs to be given")
    if logout_handler_path is None:
        raise ValueError("logout_handler_path needs to be given")
    if session_name is None:
        raise ValueError("session_name needs to be given")
    if logged_in_url is None:
        raise ValueError("logged_in_url needs to be given")
    if logged_out_url is None:
        raise ValueError("logged_out_url needs to be given")

    plugin = OpenIdIdentificationPlugin(store, 
        openid_field = openid_field,
        error_field = error_field,
        session_name = session_name,
        login_form_url = login_form_url,
        login_handler_path = login_handler_path,
        logout_handler_path = logout_handler_path,
        store_file_path = store_file_path,
        logged_in_url = logged_in_url,
        logged_out_url = logged_out_url,
        came_from_field = came_from_field,
        rememberer_name = rememberer_name,
        sql_associations_table = sql_associations_table,
        sql_nonces_table = sql_nonces_table,
        sql_connstring = sql_connstring
        )
    return plugin



repoze/who/plugins/openid/identification.py
import cgi
import urlparse
import cgitb
import sys
from zope.interface import implements

from repoze.who.interfaces import IChallenger
from repoze.who.interfaces import IIdentifier
from repoze.who.interfaces import IAuthenticator

from webob import Request, Response

import openid
from openid.store import memstore, filestore, sqlstore
from openid.consumer import consumer
from openid.oidutil import appendArgs
from openid.cryptutil import randomString
from openid.fetchers import setDefaultFetcher, Urllib2Fetcher
from openid.extensions import pape, sreg


class OpenIdIdentificationPlugin(object):
    """The repoze.who OpenID plugin
    
    This class contains 3 plugin types and is thus implementing
    IIdentifier, IChallenger and IAuthenticator.
    (check the `repoze.who documentation <http://static.repoze.org/bfgdocs/>`_
    for what all these plugin types do.)
    
    """


    implements(IChallenger, IIdentifier, IAuthenticator)

    def __init__(self, store, 
                    openid_field, 
                    error_field = '',
                    store_file_path='',
                    session_name = '',
                    login_handler_path = '',
                    logout_handler_path = '',
                    login_form_url = '',
                    logged_in_url = '',
                    logged_out_url = '',
                    came_from_field = '',
                    rememberer_name = '',
                    sql_associations_table = '',
                    sql_nonces_table = '',
                    sql_connstring = '',
                    md_provider_name = 'openidmd',
                    sreg_required = None,
                    sreg_optional = None,
                    ):

        self.rememberer_name = rememberer_name
        self.login_handler_path = login_handler_path
        self.logout_handler_path = logout_handler_path
        self.login_form_url = login_form_url
        self.session_name = session_name
        self.error_field = error_field
        self.came_from_field = came_from_field
        self.logged_out_url = logged_out_url
        self.logged_in_url = logged_in_url
        
        # for the SQL store
        self.sql_associations_table = sql_associations_table
        self.sql_nonces_table = sql_nonces_table
        self.sql_connstring = sql_connstring
        
        self.md_provider_name = md_provider_name
        self.sreg_required = sreg_required or []
        self.sreg_optional = sreg_optional or ['email','fullname', 'nickname']
        
        # set up the store
        if store==u"file":
            self.store = filestore.FileOpenIDStore(store_file_path)
        elif store==u"mem":
            self.store = memstore.MemoryStore()
        elif store==u"sql":
            # TODO: This does not work as we need a connection, not a string
            self.store = sqlstore.SQLStore(sql_connstring, sql_associations_table, sql_connstring)
        self.openid_field = openid_field
            
    def _get_rememberer(self, environ):
        rememberer = environ['repoze.who.plugins'][self.rememberer_name]
        return rememberer
    def _get_md_provider(self, environ):
        md_provider = environ['repoze.who.plugins'].get(self.md_provider_name)
        return md_provider

    def get_consumer(self,environ):
        session = environ.get(self.session_name,{})
        return consumer.Consumer(session,self.store)
        
    def redirect_to_logged_in(self, environ):
        """redirect to came_from or standard page if login was successful"""
        request = Request(environ)
        came_from = request.params.get(self.came_from_field,'')
        if came_from!='':
            url = came_from
        else:
            url = self.logged_in_url
        res = Response()
        res.status = 302
        res.location = url
        environ['repoze.who.application'] = res    

    # IIdentifier
    def identify(self, environ):
        """this method is called when a request is incoming.

        After the challenge has been called we might get here a response
        from an openid provider.

        """

        request = Request(environ)
        
        # first test for logout as we then don't need the rest
        if request.path == self.logout_handler_path:
            res = Response()
            # set forget headers
            for a,v in self.forget(environ,{}):
                res.headers.add(a,v)
            res.status = 302
            res.location = self.logged_out_url
            environ['repoze.who.application'] = res
            return {}

        identity = {}

        # first we check we are actually on the URL which is supposed to be the
        # url to return to (login_handler_path in configuration)
        # this URL is used for both: the answer for the login form and
        # when the openid provider redirects the user back.
        if request.path == self.login_handler_path:

        # in the case we are coming from the login form we should have 
        # an openid in here the user entered
            open_id = request.params.get(self.openid_field, None)
            log = environ['repoze.who.logger']
            log.debug('checking openid results for : %s ' %open_id)
            
            if open_id is not None:
                open_id = open_id.strip()
            
            # we don't do anything with the openid we found ourselves but we put it in here
            # to tell the challenge plugin to initiate the challenge
            identity['repoze.whoplugins.openid.openid'] = environ['repoze.whoplugins.openid.openid'] = open_id
        
            # this part now is for the case when the openid provider redirects
            # the user back. We should find some openid specific fields in the request.
            mode=request.params.get("openid.mode", None)
            if mode=="id_res":
                oidconsumer = self.get_consumer(environ)
                info = oidconsumer.complete(request.params, request.url)

                if info.status == consumer.SUCCESS:
                    log.info('openid request successful for : %s ' %open_id)
                    
                    display_identifier = info.identity_url
                    
                    # remove this so that the challenger is not triggered again
                    del environ['repoze.whoplugins.openid.openid']
                    
                    # store the id for the authenticator
                    identity['repoze.who.plugins.openid.userid'] = display_identifier
                    # store the user metadata...
                    try:
                        sreg_resp = sreg.SRegResponse.fromSuccessResponse(info)
                    except AttributeError, err:
                        log.warn( "Failure during SReg parsing: %s"%( err,) )
                        sreg_resp = None
                    if sreg_resp:
                        environ['repoze.who.logger'].debug("User info received: %s", sreg_resp.data)
                        user_data = dict()
                        for field in self.sreg_required + self.sreg_optional:
                            sreg_val = sreg_resp.get(field)
                            if sreg_val:
                                user_data[field] = sreg_val
                        if user_data:
                            md = self._get_md_provider( environ )
                            if md:
                                if not md.register_user( 
                                    display_identifier, user_data 
                                ):
                                    log.error( "Unable to register user" )
                                    return None
                            else:
                                log.warn( "No metadata provider %s found"%(
                                    self.md_provider_name,
                                ))
                    else:
                        log.warn(
                            "No user metadata received!"
                        )

                    # now redirect to came_from or the success page
                    self.redirect_to_logged_in(environ)
                    return identity
                    
                # TODO: Do we have to check for more failures and such?
                # 
            elif mode=="cancel":
                # cancel is a negative assertion in the OpenID protocol,
                # which means the user did not authorize correctly.
                environ['repoze.whoplugins.openid.error'] = 'OpenID authentication failed.'
                pass
        return identity

    # IIdentifier
    def remember(self, environ, identity):
        """remember the openid in the session we have anyway"""
        rememberer = self._get_rememberer(environ)
        r = rememberer.remember(environ, identity)
        return r

    # IIdentifier
    def forget(self, environ, identity):
        """forget about the authentication again"""
        rememberer = self._get_rememberer(environ)
        return rememberer.forget(environ, identity)

    # IChallenge
    def challenge(self, environ, status, app_headers, forget_headers):
        """the challenge method is called when the ``IChallengeDecider``
        in ``classifiers.py`` returns ``True``. This is the case for either a 
        ``401`` response from the client or if the key 
        ``repoze.whoplugins.openid.openidrepoze.whoplugins.openid.openid``
        is present in the WSGI environment.
        The name of this key can be adjusted via the ``openid_field`` configuration
        directive.

        The latter is the case when we are coming from the login page where the
        user entered the openid to use.

        ``401`` can come back in any case and then we simply redirect to the login
        form which is configured in the who configuration as ``login_form_url``.

        TODO: make the environment key to check also configurable in the challenge_decider.

        For the OpenID flow check `the OpenID library documentation 
        <http://openidenabled.com/files/python-openid/docs/2.2.1/openid.consumer.consumer-module.html>`_

        """

        request = Request(environ)
        
        # check for the field present, if not redirect to login_form
        if not request.params.has_key(self.openid_field):
            # redirect to login_form
            res = Response()
            res.status = 302
            res.location = self.login_form_url+"?%s=%s" %(self.came_from_field, request.url)
            return res

        # now we have an openid from the user in the request 
        openid_url = request.params[self.openid_field]
        environ['repoze.who.logger'].debug('starting openid request for : %s ' %openid_url)       

        try:
        # we create a new Consumer and start the discovery process for the URL given
        # in the library openid_request is called auth_req btw.
            openid_request = self.get_consumer(environ).begin(openid_url)
        except consumer.DiscoveryFailure, exc:
            # eventually no openid server could be found
            environ[self.error_field] = 'Error in discovery: %s' %exc[0]
            environ['repoze.who.logger'].info('Error in discovery: %s ' %exc[0])     
            return self._redirect_to_loginform(environ)
        except KeyError, exc:
            # TODO: when does that happen, why does plone.openid use "pass" here?
            environ[self.error_field] = 'Error in discovery: %s' %exc[0]
            environ['repoze.who.logger'].info('Error in discovery: %s ' %exc[0])
            return self._redirect_to_loginform(environ)
            return None
        # not sure this can still happen but we are making sure.
        # should actually been handled by the DiscoveryFailure exception above
        if openid_request is None:
            environ[self.error_field] = 'No OpenID services found for %s' %openid_url
            environ['repoze.who.logger'].info('No OpenID services found for: %s ' %openid_url)
            return self._redirect_to_loginform(environ)
       
        # we have to tell the openid provider where to send the user after login
        # so we need to compute this from our path and application URL
        # we simply use the URL we are at right now (which is the form)
        # this will be captured by the repoze.who identification plugin later on
        # it will check if some valid openid response is coming back
        # trust_root is the URL (realm) which will be presented to the user
        # in the login process and should be your applications url
        # TODO: make this configurable?
        # return_to is the actual URL to be used for returning to this app
        return_to = request.path_url # we return to this URL here
        trust_root = request.application_url
        environ['repoze.who.logger'].debug('setting return_to URL to : %s ' %return_to)
        
        # TODO: usually you should check openid_request.shouldSendRedirect()
        # but this might say you have to use a form redirect and I don't get why
        # so we do the same as plone.openid and ignore it.
        
        # Request additional information (optional here, could require fields as well)...
        if self.sreg_optional or self.sreg_required:
            openid_request.addExtension(
                sreg.SRegRequest(
                    required = self.sreg_required,
                    optional = self.sreg_optional,
                )
            )

        # TODO: we might also want to give the application some way of adding
        # extensions to this message.
        redirect_url = openid_request.redirectURL(trust_root, return_to) 
        # # , immediate=False)
        res = Response()
        res.status = 302
        res.location = redirect_url
        environ['repoze.who.logger'].debug('redirecting to : %s ' %redirect_url)

        # now it's redirecting and might come back via the identify() method
        # from the openid provider once the user logged in there.
        return res
        
    def _redirect_to_loginform(self, environ={}):
        """redirect the user to the login form"""
        res = Response()
        res.status = 302
        q=''
        ef = environ.get(self.error_field, None)
        if ef is not None:
                q='?%s=%s' %(self.error_field, ef)
        res.location = self.login_form_url+q
        return res
        
                
    # IAuthenticator
    def authenticate(self, environ, identity):
        """dummy authenticator
        
        This takes the openid found and uses it as the userid. Normally you would want
        to take the openid and search a user for it to map maybe multiple openids to a user.
        This means for you to simply implement something similar to this. 
        
        """
        if identity.has_key("repoze.who.plugins.openid.userid"):
                environ['repoze.who.logger'].info('authenticated : %s ' %identity['repoze.who.plugins.openid.userid'])
                return identity.get('repoze.who.plugins.openid.userid')


    def __repr__(self):
        return '<%s %s>' % (self.__class__.__name__, id(self))

